### PR TITLE
port Docker updater improvements from Azure DevOps

### PR DESCRIPTION
**Pull request by brettfo**
As part of the work inside Azure DevOps to improve dependabot (see #8179) there were some updates to the Docker updater that needed to be split out into their own PR, specifically:

- Allow docker version prefixes to contain underscores
- Add support for digests when updating helm charts

(cc @brbayes-msft)

### CODE REVIEW

**Comment by abdulapopoola** [comment]
Thank you so much for this! Would it be possible to have some specs to avoid regressions in the future please?

File: docker/lib/dependabot/docker/tag.rb [hunk 1]
```diff
@@ -9,8 +9,8 @@
       WORDS_WITH_BUILD = /(?:(?:-[a-z]+)+-[0-9]+)+/
       VERSION_REGEX = /v?(?<version>[0-9]+(?:\.[0-9]+)*(?:_[0-9]+|\.[a-z0-9]+|#{WORDS_WITH_BUILD}|-(?:kb)?[0-9]+)*)/i
       VERSION_WITH_SFX = /^#{VERSION_REGEX}(?<suffix>-[a-z][a-z0-9.\-]*)?$/i
-      VERSION_WITH_PFX = /^(?<prefix>[a-z][a-z0-9.\-]*-)?#{VERSION_REGEX}$/i
-      VERSION_WITH_PFX_AND_SFX = /^(?<prefix>[a-z\-]+-)?#{VERSION_REGEX}(?<suffix>-[a-z\-]+)?$/i
+      VERSION_WITH_PFX = /^(?<prefix>[a-z][a-z0-9.\-_]*-)?#{VERSION_REGEX}$/i
```
[the remark this answers is not in the document]
Comment (ryanbrandenburg) [reply docker/lib/dependabot/docker/tag.rb:12]: @abdulapopoola I'm not the greatest regex wizard but I think this would be a pre-existing condition (IE, this was not caused by our change to this REGEX but was already "inefficient" and just somehow grandfather-ed in). Do you have any advise for resolving this given that that's the case.

[the remark this answers is not in the document]
Comment (abdulapopoola) [reply docker/lib/dependabot/docker/tag.rb:12]: As a team, I think we accept RE as necessary and that codeQL flagging might be too sensitive; tagging @deivid-rodriguez and @jakecoffman to get their thoughts.

[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply docker/lib/dependabot/docker/tag.rb:12]: I think it's fine for CodeQL to flag these but I don't like it that it flags issues on _any_ changed lines, even if they were present also before the changes. It places unnecessary burden on contributions like here.

Anyways, I believe it's fine to ignore.

Still, I had a look and I think this change may fix it, and it does not seem to break any specs:

```diff
diff --git a/docker/lib/dependabot/docker/tag.rb b/docker/lib/dependabot/docker/tag.rb
index 65c21e267..8eb805614 100644
--- a/docker/lib/dependabot/docker/tag.rb
+++ b/docker/lib/dependabot/docker/tag.rb
@@ -8,7 +8,7 @@ module Dependabot
     class Tag
       WORDS_WITH_BUILD = /(?:(?:-[a-z]+)+-[0-9]+)+/
       VERSION_REGEX = /v?(?<version>[0-9]+(?:\.[0-9]+)*(?:_[0-9]+|\.[a-z0-9]+|#{WORDS_WITH_BUILD}|-(?:kb)?[0-9]+)*)/i
-      VERSION_WITH_SFX = /^#{VERSION_REGEX}(?<suffix>-[a-z][a-z0-9.\-]*)?$/i
+      VERSION_WITH_SFX = /^#{VERSION_REGEX}(?<suffix>-[a-z][a-z0-9.]*)?$/i
       VERSION_WITH_PFX = /^(?<prefix>[a-z][a-z0-9.\-]*-)?#{VERSION_REGEX}$/i
       VERSION_WITH_PFX_AND_SFX = /^(?<prefix>[a-z\-]+-)?#{VERSION_REGEX}(?<suffix>-[a-z\-]+)?$/i
       NAME_WITH_VERSION =
```

Just in case we want to change that separately.

File: docker/spec/dependabot/docker/file_parser_spec.rb [hunk 2]
```diff
@@ -462,6 +462,34 @@
...
+              registry: "registry-host.io:5000",
+              tag: "17.04_"
+            }
+          }]
+        end
+
+        it "has the right details" do
+          expect(dependency).to be_a(Dependabot::Dependency)
+          expect(dependency.name).to eq("myreg/ubuntu")
+          expect(dependency.version).to eq("17.04_")
```
Comment (deivid-rodriguez) [inline docker/spec/dependabot/docker/file_parser_spec.rb:487]: Is it this a real example, or just a simplified version for the sake of testing? Do you have an example of a realworld image tagged like this?

So far we've been trying to support to most widely used versioning schemes out there since there's no official spec for versioning docker tags. But arbitrarily putting an underscore after the version number does not make a lot of sense to me.

Comment (ryanbrandenburg) [reply docker/spec/dependabot/docker/file_parser_spec.rb:487]: This is not a real example. The version we're thinking of is non-public, but it's a docker container for which the tag is shaped something like `someRepo_19700101.4`. I can replace the above with that now that I know what shape we're imagining.

**Comment by ryanbrandenburg** [comment]
This should be ready for re-review.

**Comment by deivid-rodriguez** [comment]
Would it be possible to split this PR into two commits, one for the helm charts feature, and another way for the detection of docker images with underscored suffixes? That'd help with the review!

File: docker/spec/dependabot/docker/file_parser_spec.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply docker/spec/dependabot/docker/file_parser_spec.rb]: So, just to confirm, you need to support underscores as suffixes to the version, not as prefixes as originally stated, correct?

**Comment by ryanbrandenburg** [comment]
I have rebased the commits as per your request.

File: docker/lib/dependabot/docker/file_updater.rb [hunk 3]
```diff
@@ -158,7 +158,7 @@ def update_helm(file, content)
         old_tags.each do |old_tag|
           old_tag_regex = /^\s+(?:-\s)?(?:tag|version):\s+["']?#{old_tag}["']?(?=\s|$)/
           modified_content = modified_content.gsub(old_tag_regex) do |old_img_tag|
-            old_img_tag.gsub(old_tag.to_s, new_yaml_tag(file).to_s)
+            old_img_tag.gsub(old_tag.to_s, new_helm_tag(file).to_s)
```
Comment (deivid-rodriguez) [inline docker/lib/dependabot/docker/file_updater.rb:161]: The old method is dead now?

File: docker/lib/dependabot/docker/file_parser.rb [hunk 4]
```diff
@@ -181,6 +193,8 @@ def parse_helm(img_hash)
           []
         end
       end
+      # rubocop:enable Metrics/PerceivedComplexity
+      # rubocop:enable Metrics/AbcSize
```
Comment (deivid-rodriguez) [inline docker/lib/dependabot/docker/file_parser.rb:197]: This got a bit out of hand, let me propose a simpler way:

```diff
-      # rubocop:disable Metrics/AbcSize
-      # rubocop:disable Metrics/PerceivedComplexity
       def parse_helm(img_hash)
         repo = img_hash.fetch("repository", nil)
         tag_value = img_hash.key?("tag") ? img_hash.fetch("tag", nil) : img_hash.fetch("version", nil)
@@ -178,22 +176,14 @@ module Dependabot
         tag = tag_details["tag"]
         digest = tag_details["digest"]
 
-        if !repo.nil? && !registry.nil? && !tag.nil? && !digest.nil?
-          ["#{registry}/#{repo}:#{tag}@sha256:#{digest}"]
-        elsif !repo.nil? && !registry.nil? && !tag.nil?
-          ["#{registry}/#{repo}:#{tag}"]
-        elsif !repo.nil? && !tag.nil? && !digest.nil?
-          ["#{repo}:#{tag}@sha256:#{digest}"]
-        elsif !repo.nil? && !tag.nil?
-          ["#{repo}:#{tag}"]
-        elsif !repo.nil?
-          [repo]
-        else
-          []
-        end
+        return [] unless repo
+        return [repo] unless tag
+
+        image = "#{repo}:#{tag}"
+        image.prepend("#{registry}/") if registry
+        image.append("@sha256:#{digest}/") if digest
+        [image]
       end
-      # rubocop:enable Metrics/PerceivedComplexity
-      # rubocop:enable Metrics/AbcSize
```

Also, I think we have no tests for parsing helm charts, not this PRs job but we should probably add some.

File: docker/spec/dependabot/docker/file_updater_spec.rb [hunk 5]
```diff
@@ -1316,6 +1358,46 @@
...
+
+      describe "the updated helmfile" do
+        subject(:updated_helmfile) do
+          updated_files.find { |f| f.name == "values.yaml" }
+        end
+
+        its(:content) { is_expected.to include "image:\n  repository: 'nginx'\n  tag: 1.14.3\n" }
+      end
+    end
+
```
Comment (deivid-rodriguez) [inline docker/spec/dependabot/docker/file_updater_spec.rb:1400]: If I completely revert the changes in `docker/lib` about updating heml charts, these specs still pass? Maybe they're not covering the right code?

**Review by deivid-rodriguez — COMMENTED** [review]
This is looking good. The changes for supporting underscore prefixes look good to me, feel free to propose separately if you want those merged straightaway. For heml charts I had some questions.

File: docker/lib/dependabot/docker/file_updater.rb (hunk 3 above)
Comment (brettfo) [reply docker/lib/dependabot/docker/file_updater.rb:161]: Correct, that method is no longer used, I'll remove it shortly.

File: docker/lib/dependabot/docker/file_parser.rb (hunk 4 above)
Comment (brettfo) [reply docker/lib/dependabot/docker/file_parser.rb:197]: Good call!  Will update and push changes.

File: docker/spec/dependabot/docker/file_updater_spec.rb (hunk 5 above)
Comment (brettfo) [reply docker/spec/dependabot/docker/file_updater_spec.rb:1400]: I think the test name was a copy/paste error.  I've renamed the test and pushed an update.

**Comment by brettfo** [comment]
@deivid-rodriguez I've cleaned up the commits based on your feedback and rebased on the current `main`.

**Review by jakecoffman — APPROVED** [review]
